File: web-admin/src/api/journal.ts
import { api } from './http';

export interface JournalLine {
    id: string;
    account_id: string;
    description?: string;
    debit: number;
    credit: number;
}

export interface JournalEntry {
    id: string;
    transaction_number: string;
    date: string;
    description: string;
    reference?: string;
    status: 'draft' | 'posted';
    created_by?: string;
    created_at: string;
}

export interface JournalEntryDetail {
    header: JournalEntry;
    lines: JournalLine[];
}

export interface CreateJournalLineRequest {
    account_id: string;
    description?: string;
    debit: number;
    credit: number;
}

export interface CreateJournalEntryRequest {
    date: string;
    description: string;
    reference?: string;
    lines: CreateJournalLineRequest[];
}

export interface ListJournalsParams {
    page?: number;
    limit?: number;
}

export const journalApi = {
    list: async (params?: ListJournalsParams) => {
        const response = await api.get<JournalEntry[]>('/finance/journals', { params });
        return response.data;
    },

    getDetails: async (id: string) => {
        const response = await api.get<JournalEntryDetail>(`/finance/journals/${id}`);
        return response.data;
    },

    create: async (data: CreateJournalEntryRequest) => {
        const response = await api.post<JournalEntryDetail>('/finance/journals', data);
        return response.data;
    },
};

File: crates/finance/src/domain/entities/journal.rs
//! Pure Journal Entry Entities (3R.1.1-002)
//!
//! Owned strictly by `crates/finance`. Contains NO sqlx persistence attributes.

use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum JournalStatus {
    Draft,
    Posted,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JournalEntry {
    pub id: Uuid,
    pub transaction_number: String,
    pub date: NaiveDate,
    pub description: String,
    pub reference: Option<String>,
    pub status: JournalStatus,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JournalLine {
    pub id: Uuid,
    pub journal_entry_id: Uuid,
    pub account_id: Uuid,
    pub description: Option<String>,
    pub debit: Decimal,
    pub credit: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JournalEntryDetail {
    #[serde(flatten)]
    pub header: JournalEntry,
    pub lines: Vec<JournalLine>,
}

#[derive(Debug, Deserialize)]
pub struct CreateJournalLineRequest {
    pub account_id: Uuid,
    pub description: Option<String>,
    pub debit: Decimal,
    pub credit: Decimal,
}

#[derive(Debug, Deserialize)]
pub struct CreateJournalEntryRequest {
    pub date: NaiveDate,
    pub description: String,
    pub reference: Option<String>,
    pub lines: Vec<CreateJournalLineRequest>,
}
